### PR TITLE
Added support for data() call for C++11 onwards

### PR DESCRIPTION
**Pull request by rishitc**
Added preprocessor check for C++ standard used for compilation. If a C++11 or after standard is detected then the ```data()``` method is used to get the pointer to the first element in the array which is used internally by the vector.

### CODE REVIEW

**Comment by rishitc** [comment]
Hi,
I've just pushed another pull request and it seems to have failed the tests.
I'm not sure what has caused the error this time. I added the preprocessor checks as discussed, to all the functions where I made the changes and I checked the preprocessor outputs locally as well, to confirm if the implementation is correctly selected based on the C++ standard.

Any inputs on why this would be causing code tests to fail, would be really helpful!

Thanks

File: common/vector.hpp [hunk 1]
```diff
@@ -34,15 +34,64 @@ namespace acommon
...
+      #endif
+    }
+    T * data(int pos)
+    {
+      #if __cplusplus >= 201103L
+        return this->data() + pos;
+      #else
+        return &*this->begin() + pos;
+      #endif
+    }
```
Comment (kevina) [inline common/vector.hpp:52]: Please just make this:
```
T * data(int pos) {return data() + pos;}
```

File: common/vector.hpp [hunk 2]
```diff
@@ -34,15 +34,64 @@ namespace acommon
...
-    T * data(int pos) {return &*this->begin() + pos;}
-    T * data_end() {return &*this->begin() + this->size();}
+    T * data()
+    {
+      #if __cplusplus >= 201103L
+        return this->data();
+      #else
+        return &*this->begin();
+      #endif
+    }
```
Comment (kevina) [inline common/vector.hpp:44]: This is wrong, it is calling itself.

All you need to do is:
```
#if __cplusplus < 201103L
T * data() {return &*this->begin();}
#endif
```

**Review by kevina — CHANGES_REQUESTED** [review]
The data() method is calling itself.

All other methods should just call the data() method, see comment for data(int pos).

File: common/vector.hpp (hunk 1 above)
Comment (rishitc) [reply common/vector.hpp:52]: Hi,
I'm extremely sorry about that; this was quite clumsy of me to overlook the function signature.
I will make the changes and submit another pull request.

Thanks

**Comment by rishitc** [comment]
I tried the recommended changes in #609 but the tests failed again. I've quoted the description of the changes I made in that pull request, here:

> I made the recommended changes but the tests still failed. I then even tried merging the ```&*begin()``` calls to ```data()```.
> The vector class on compiling locally on my system, gave no errors.
>
> Any inputs of why the tests failed would be really helpful!
>
> I did check CircleCI, but the error messages reported are not useful. The website also says some changes need to be made to the config file to save the test results.
>
>Thanks.

**Comment by kevina** [comment]
Please push your changes to this p.r.

**Comment by rishitc** [comment]
I have pushed my changes to this PR. As mentioned, I'll push all my further changes regarding this issue to this PR, as well.

Thanks.

**Comment by kevina** [comment]
Thank you for your contribution.  The tests need to be fixed.  I will try to get to that within the next couple of days.

Your code looks correct, but there are some minor nitpicks.  To save time, once I fix the tests I will go ahead and redo your commit and push it, closing this request.

**Comment by rishitc** [comment]
Hi,
Thank you so much for you inputs and support, while I was working on this pull request. It was really fun working on this pull request. I'll take a look at the issues list of the project and try solving some of the issues there in the coming weeks.
I guess should start that, once the tests are fixed?

> but there are some minor nitpicks

What nitpicks are you referring to? I would love to know, what they are, so that I can learn about them and not make those mistakes for any future PRs I submit, for other issues.

Thanks.

**Comment by kevina** [comment]
> What nitpicks are you referring to?

You reordered the methods making it difficult to see the changes made.  For such a small change it is faster for me to just fix it rather than lots of back and forth.  I will be happy to answer any questions you have once I push the changes.

**Comment by kevina** [comment]
Please don't use this PR for unrelated conversion and instead open a new issue.

**Comment by rishitc** [comment]
Sorry, about that, I've deleted my message.

**Pull request closed without merging**